Add master-slave replication support

sameersbn opened the pull request:
Example,

Start master:

```bash
docker run -it --rm --name mariadb-master \
  -e 'MARIADB_SERVER_ID=1' \
  -e 'MARIADB_USER=root' -e 'MARIADB_PASSWORD=password' \
  -e 'MARIADB_DATABASE=newdatabase' \
  -e 'MARIADB_REPLICATION_MODE=master' \
  -e 'REPLICATION_USER=slave' -e 'MARIADB_REPLICATION_PASSWORD=slavepassword' \
  bitnami/mariadb
```

Start Slave:

```bash
docker run -it --rm --name mariadb-slave \
  --link mariadb-master:master \
  -e 'MARIADB_SERVER_ID=2'
  -e 'MARIADB_REPLICATION_MODE=slave' \
  bitnami/mariadb
```

bitnami-container-builds commented:
Can one of the admins verify this patch?

sameersbn commented:
Work in progress. For your review only.
Do not merge.

sameersbn commented:
> Is it needed to add the log-bin option in the slave as well?

May not be required. I will test it and remove it if not required. Thanks for pointing it out.

> Also, regarding the binary log name, what about renaming it to include the -log on it. Right now It might be confusing with the binary executable name.

We can change it. Wherever I looked, everyone seemed to be using `mysql-bin`. But yes, its just a name. We can change it.

> Why did you choose to set up the binary log format to ROW and no Mixed for example?

ROW based replication is regarded as the safest form of replication so I think that is a good default to ensure the integrity of users data. Additionally `STATEMENT` based replication requires that `USE database` instruction is specified. This means that a statement like `INSERT INTO mydb.users(name) VALUES ('Marko');` will not be replicated if `USE mydb` was not done prior to the `INSERT` statement.

Users should be able to switch to `STATEMENT` or `MIXED` using command line arguments. I just think that `ROW` is the safer default. I am not very clear on `MIXED` mode, some investigation may be required. 
 
https://dev.mysql.com/doc/refman/5.1/en/replication-sbr-rbr.html

> In case that we pass the MARIADB_DATABASE to the master container, shouldn't we restrict the replication access to only that database?

Yes, thanks for pointing it out. I will restrict the replication user to the MARIADB_DATABASE and verify that it is ok.

> This is for setting up the initial state of the database we want to replicate right?

That is correct. Before the slave can be started we have to set it to the initializes state of the database on the master.

> What if during the dump there are new changes in the master database, will these get picked by the dump and the start slave command?

The `--master-dump` argument on `mysqldump` automatically turns on the `--lock-all-tables` option. This should ensure that no writes occur while the dump is being generated. refer https://dev.mysql.com/doc/refman/5.1/en/mysqldump.html#option_mysqldump_master-data

> Please keep in mind that my knowledge about setting up clustering in Mariadb is very limited so some of my comments might not make sense to you :)

Well... you are not alone. This is new to me as well :)

sameersbn commented:
Please note I have updated the PR with the suggested changes and some fixes. Right now the PR has multiple commits to allow you to easily go through the changes. When we are in a ready to merge state I will squash all the commits.

sameersbn commented:
@migmartri I have updated the PR so that there is zero downtime of the master when a slave is started.

migmartri commented on bitnami-utils-custom.sh:
```diff
@@ -1,6 +1,15 @@
 
+case "$REPLICATION_MODE" in
+  master )
+    PROGRAM_OPTIONS+=" --server-id=${SERVER_ID:-$RANDOM} --binlog-format=ROW --log-bin=mysql-bin --innodb_flush_log_at_trx_commit=1 --sync-binlog=1"
```
Actually --innodb_flush_log_at_trx_commit=1 is set by default to 1 right?

sameersbn replied:
@migmartri its true that `--innodb_flush_log_at_trx_commit=1` is the default. I thought it made sense to explicitly specify that option to be clear that it is one of the options required to get the replication working correctly. Users should be able to override that option because user specified options are appended to the command line.

migmartri commented on bitnami-utils-custom.sh:
```diff
@@ -40,12 +49,45 @@ create_mysql_user() {
+
+      echo "==> Creating a data snapshot..."
+      mysqldump -u$MASTER_USER ${MASTER_PASSWORD:+-p$MASTER_PASSWORD} -h $MASTER_HOST \
+        --databases $MARIADB_DATABASE --skip-lock-tables --single-transaction --flush-logs --hex-blob --master-data --apply-slave-statements --comments=false | tr -d '\012' | sed -e 's/;/;\n/g' >> /tmp/init_mysql.sql
```
$MARIADB_DATABASE is not an mandatory option right? Shouldn't we dump all the databases (-A option) instead in that case?

Also, we use --master-data (equal=1) which includes the CHANGE MASTER statement. Any idea why in the other example we talked about they used --master-data=2 and then change the master manually?

sameersbn replied:
> $MARIADB_DATABASE is not an mandatory option right? Shouldn't we dump all the databases (-A option) instead in that case?

- Dumping all databases would also mean that a dump would take a longer time if the database server has many large databases and user is only interested in replicating a particular database.
- It would also need that all the databases in the dump exist at the the slave for the restore to happen since the dump will not have `CREATE DATABASE` statements.

> Also, we use --master-data (equal=1) which includes the CHANGE MASTER statement. Any idea why in the other example we talked about they used --master-data=2 and then change the master manually?

I think https://github.com/sameersbn/bitnami-docker-mariadb/commit/7139b50406520d2a961faaaa2feb5986e0b4e44d is the main reason why `--master-data=2` is used in the example. By using `--master-data=1` we are avoiding doing a `grep` to get the `LOG_FILE_POS` and `LOG_FILE_SIZE` and to avoid the situation mentioned in the referenced commit, we are setting the `MASTER_HOST` before restoring the dump that has the set `MASTER_LOG_POS` and `MASTER_LOG_SIZE` statements.

migmartri replied:
> Dumping all databases would also mean that a dump would take a longer time if the database server has many large databases and user is only interested in replicating a particular databas

I was thinking about if you pass the $MARIADB_DATABASE dump this one, otherwise dump them all. Basically the user will say enable replication to one database or all of them.

> It would also need that all the databases in the dump exist at the the slave for the restore to happen since the dump will not have CREATE DATABASE statements and at the moment at least the bitnami/mariadb container does not allow users to create multiple databases when started.

I thought that mysqldump with the -A option adds by default the CREATE DATABASE statement.  

>  at the moment at least the bitnami/mariadb container does not allow users to create multiple databases when started.

I was thinking more about the scenario in which you have a master with multiple vms, then you enable replication and then you try to use our image to get slave replication.

In any case, this scenario might not worth working on it. But shouldn't we tell the user that it needs to provide MARIADB_DATABASE if REPLICATION_MODE=slave is there?

sameersbn replied:
> I was thinking about if you pass the $MARIADB_DATABASE dump this one, otherwise dump them all. Basically the user will say enable replication to one database or all of them.

But the `MARIADB_USER` is needed to be created on the slave  should have access to the database we want to replicate. Otherwise the slave will be un-usable because there will not be a user that can access the database. Right now the `MARIADB_USER` and `MARIADB_DATABASE` parameters specified on the command line makes sure that the user has access to the database.

If no `MARIADB_DATABASE` is specified and we are replicating all databases, then the created user will not have access to any database or else we will need to list all databases and give the user access to all the databases.

> I thought that mysqldump with the -A option adds by default the CREATE DATABASE statement.

You are right. It adds the `CREATE DATABASE` statement else the dump will not make sense. My bad.

> I was thinking more about the scenario in which you have a master with multiple vms, then you enable replication and then you try to use our image to get slave replication.
> 
> In any case, this scenario might not worth working on it. But shouldn't we tell the user that it needs to provide MARIADB_DATABASE if REPLICATION_MODE=slave is there?

Yes, we will mention in the docs that when started as a slave, the `MARIADB_DATABASE` parameter enables replication of that database only.

sameersbn commented:
the `ln -s /dev/stdout logs/mysqld.log` in the `post-install.sh` script of the MariaDB image seems to break the zero-downtime replication due to the presense of `--flush-logs` option specified in the `mysqldump` command. Mounting a volume at `/bitnami/mariadb/logs` or commenting out the line from the `post-install.sh` script seems to resolve the issue.

Steps to reproduce

Build bitnami/mariadb from the `master-slave-replication-support` git branch

Create Master:

```docker run -it --rm --name mariadb-master \
  -e 'MARIADB_USER=redmine' -e 'MARIADB_PASSWORD=password' \
  -e 'MARIADB_DATABASE=redmine_production' -e 'REPLICATION_MODE=master' \
  -e 'REPLICATION_USER=slave' -e 'REPLICATION_PASSWORD=slavepassword' \
  -v /tmp/mariadb-master/data:/bitnami/mariadb/data \
  bitnami/mariadb
```

Create Slave:

```docker run -it --rm --name mariadb-slave --link mariadb-master:mariadb-master \
  -e 'MARIADB_USER=redmine' -e 'MARIADB_PASSWORD=password' \
  -e 'MARIADB_DATABASE=redmine_production' -e 'REPLICATION_MODE=slave' \
  bitnami/mariadb
```

Stop Master and Slave:

```docker stop mariadb-slave mariadb-master
````

Restart Master from previous command and you will see that the master will error out.

Here is the error log

```150730 14:16:14 [Note] mysqld.bin (mysqld 5.5.44-MariaDB-log) starting as process 1 ...
150730 14:16:14 InnoDB: The InnoDB memory heap is disabled
150730 14:16:14 InnoDB: Mutexes and rw_locks use GCC atomic builtins
150730 14:16:14 InnoDB: Compressed tables use zlib 1.2.3
150730 14:16:14 InnoDB: Using Linux native AIO
150730 14:16:14 InnoDB: Initializing buffer pool, size = 128.0M
150730 14:16:14 InnoDB: Completed initialization of buffer pool
150730 14:16:14 InnoDB: highest supported file format is Barracuda.
150730 14:16:14  InnoDB: Waiting for the background threads to start
150730 14:16:15 Percona XtraDB (http://www.percona.com) 5.5.43-MariaDB-37.2 started; log sequence number 1597945
150730 14:16:15 [Note] Plugin 'FEEDBACK' is disabled.
150730 14:16:15 [ERROR] mysqld.bin: File '150730 14:16:10 [Note] mysqld.bin: Normal shutdown' not found (Errcode: 2)
150730 14:16:15 [ERROR] Failed to open log (file '150730 14:16:10 [Note] mysqld.bin: Normal shutdown', errno 2)
150730 14:16:15 [ERROR] Could not open log file
150730 14:16:15 [ERROR] Can't init tc log
150730 14:16:15 [ERROR] Aborting

150730 14:16:15  InnoDB: Starting shutdown...
150730 14:16:16  InnoDB: Shutdown completed; log sequence number 1597945
150730 14:16:16 [Note] mysqld.bin: Shutdown complete
```

As mentioned above I found this to be issue issue when `--flush-logs` is added to the `mysqldump` command that is executed when the slave is started and this seems to somehow cause an issue because `logs/mysqld.log` is symlinked to `/dev/stdout`.

migmartri commented:
@sameersbn 

I have noticed that the mysql-bin.index includes some wrong extra lines which seems to be output logs:

This happens also without mounting the data directory in the master.

```
root@cd04ff1236ed:/opt/bitnami/mysql/data# cat mysql-bin.index
./mysql-bin.000001
./mysql-bin.000002
150730 18:16:43 [Note] mysqld.bin: Normal shutdown

150730 18:16:43 [Note] Event Scheduler: Purging the queue. 0 events
150730 18:16:43  InnoDB: Starting shutdown...
150730 18:16:43  InnoDB: Shutdown completed; log sequence number 1597945
150730 18:16:44 [Note] mysqld.bin: Shutdown complete 
```

migmartri commented:
I have been tracking down the issue and this is what I have found out:

1 - We start the master, 

    root@37f5d4d95227:/# cat /opt/bitnami/mysql/data/mysql-bin.index
    ./mysql-bin.000001

2 - We start the slave

    root@37f5d4d95227:/# cat /opt/bitnami/mysql/data/mysql-bin.index
    ./mysql-bin.000001
    ./mysql-bin.000002

3 - We stop the slave

    docker stop mariadb-slave

    root@37f5d4d95227:/# cat /opt/bitnami/mysql/data/mysql-bin.index
    ./mysql-bin.000001
    ./mysql-bin.000002

4 - We start the slave

  docker start mariadb-slave

    root@37f5d4d95227:/# cat /opt/bitnami/mysql/data/mysql-bin.index
    ./mysql-bin.000001
    ./mysql-bin.000002
    150730 18:28:58 [Warning] IP address '172.17.0.29' could not be resolved: Name or service not known  

I can confirm that the trash line append happens when you start the new slave.

sameersbn commented:
@migmartri I see the same. However when the `logs/mysqld.log` symlink to `/dev/stdout` is not present, this issue does not occur.

migmartri commented:
@sameersbn it seems that we get the warning every time we connect to a mariadb container from another one regardless of the master-slave setup.

I do not know why that warning message get's mixed in in the index file but it seems that it might be possible [to remove it](http://serverfault.com/questions/393862/mysql-warning-ip-address-could-not-be-resolved). 

I think that removing the warning message will be a good thing anyways because of the default docker network isolation. 

What do you think @sameersbn ?

sameersbn commented:
@migmartri it will get rid of the one message ending up in the dump. what in case of other messages that we may have not encountered yet. However I will give it a try and let you know the results.

sameersbn commented:
Was checking this.. the lines from your comment https://github.com/bitnami/bitnami-docker-mariadb/pull/17#issuecomment-126424033 end up in the `mysql-bin.index` file when the slave connects. The `skip_name_resolve` rightly has no effect on those lines.

What if we remove the stdout symlink and instead use the `wait_and_tail_logs` function?

migmartri commented:
> What if we remove the stdout symlink and instead use the wait_and_tail_logs function?

wait_and_tail logs was a function that we were thinking of deprecating it eventually since it creates another process inside the container but in the meantime and since we do not have a better option I think you are right and we should use it in this case.

So skip_name_resolve does not stop showing the message even in regular container links? You can reproduce getting this message creating a mysql server, then another container that is linked to this server and then disconnect it.

Thanks,
Miguel

sameersbn commented:
> wait_and_tail logs was a function that we were thinking of deprecating it eventually since it creates another process inside the container but in the meantime and since we do not have a better option I think you are right and we should use it in this case.

I completely agree. I am not a big fan of `wait_and_tail_logs` either. In the current state, it seems the only viable option.

sameersbn commented:
> So skip_name_resolve does not stop showing the message even in regular container links?

With regards to this, with the latest mariadb image I don't see these such `150730 18:28:58 [Warning] IP address '172.17.0.29' could not be resolved: Name or service not known` messages in the logs regardless of the state of `skip_name_resolve` option. I user the instructions from https://github.com/bitnami/bitnami-docker-mariadb#connecting-a-mysql-client-container-to-the-mariadb-server-container

sameersbn commented:
updated the PR to replace the symlink with `wait_and_tail_logs`. Will add the required tests and update the README before removing the **WIP** label.

migmartri commented:
Hi @sameersbn, 

I am able to reproduce the issue I do:

docker run --rm --name mariadb-server bitnami/mariadb

And then:

docker run -it --rm --link mariadb-server:mariadb bitnami/mariadb mysql -h mariadb -u root

This produces in the server logs the message.

sameersbn commented:
@migmartri tried using your commands. I cannot see any messages when client connects to the server. After the following initial log I don't see anything new messages when the client container connects.

```
150909 14:26:42 [Note] mysqld.bin (mysqld 5.5.45-MariaDB) starting as process 1 ...
150909 14:26:42 InnoDB: The InnoDB memory heap is disabled
150909 14:26:42 InnoDB: Mutexes and rw_locks use GCC atomic builtins
150909 14:26:42 InnoDB: Compressed tables use zlib 1.2.3
150909 14:26:42 InnoDB: Using Linux native AIO
150909 14:26:42 InnoDB: Initializing buffer pool, size = 128.0M
150909 14:26:42 InnoDB: Completed initialization of buffer pool
InnoDB: The first specified data file ./ibdata1 did not exist:
InnoDB: a new database to be created!
150909 14:26:42  InnoDB: Setting file ./ibdata1 size to 10 MB
InnoDB: Database physically writes the file full: wait...
150909 14:26:42  InnoDB: Log file ./ib_logfile0 did not exist: new to be created
InnoDB: Setting log file ./ib_logfile0 size to 5 MB
InnoDB: Database physically writes the file full: wait...
150909 14:26:42  InnoDB: Log file ./ib_logfile1 did not exist: new to be created
InnoDB: Setting log file ./ib_logfile1 size to 5 MB
InnoDB: Database physically writes the file full: wait...
InnoDB: Doublewrite buffer not found: creating new
InnoDB: Doublewrite buffer created
InnoDB: 127 rollback segment(s) active.
InnoDB: Creating foreign key constraint system tables
InnoDB: Foreign key constraint system tables created
150909 14:26:42  InnoDB: Waiting for the background threads to start

150909 14:26:43 Percona XtraDB (http://www.percona.com) 5.5.44-MariaDB-37.3 started; log sequence number 0
150909 14:26:43 [Note] Plugin 'FEEDBACK' is disabled.
150909 14:26:43 [Note] Server socket created on IP: '0.0.0.0'.
150909 14:26:43 [Note] Event Scheduler: Loaded 0 events
150909 14:26:43 [Note] mysqld.bin: ready for connections.
Version: '5.5.45-MariaDB'  socket: '/opt/bitnami/mysql/tmp/mysql.sock'  port: 3306  MariaDB Server
```
Just to be sure my image has `BITNAMI_APP_VERSION=5.5.45-0` and I am using the most recent `ubuntu-debootstrap:14.04` baseimage which was release about 2 weeks ago. Maybe some change in the base image that causes the message to go away. Probably something in the resolver.

sameersbn commented:
Added the tests and updated the docs.
I have remove the WIP label to indicate that it is ready to be merged.
Please review the PR and let me know if any changes are required.

sameersbn commented:
- Added the `MARIADB_` prefix to the replication parameters
- Added support for fetching `MARIADB_USER`, `MARIADB_PASSWORD` and `MARIADB_DATABASE` parameters (from master) in slave when not specified.

Status: Ready for merge

sameersbn commented:
- Added the `MARIADB_` prefix to the replication parameters
- Added support for fetching `MARIADB_USER`, `MARIADB_PASSWORD` and `MARIADB_DATABASE` parameters (from master) in slave when not specified.

Status: Ready for merge

migmartri commented:
test this please
